experimental/alp/lib/Transforms/modulo_scheduling_pass.cpp

```diff
@@ -0,0 +1,136 @@
+//===- extract_kernel_pass.cpp - Pipeline the kernel -===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+#include "alp/Transforms/Passes.h"
+#include "alp/Transforms/PassDetail.h"
+
+#include "mlir/Pass/Pass.h"
+#include "mlir/Pass/PassManager.h"
+#include "mlir/Transforms/Passes.h"
+#include <unordered_map>
+
+#include "mlir/Dialect/SCF/SCF.h"
+#include "mlir/Dialect/SCF/Transforms.h"
+#include "mlir/Dialect/SCF/Utils.h"
+#include "mlir/Dialect/Vector/VectorOps.h"
+#include "mlir/Transforms/LoopUtils.h"
+#include "mlir/Transforms/GreedyPatternRewriteDriver.h"
+
+
+using namespace mlir;
+
+namespace{
+  struct ModuloSchedulingPass : public ModuloSchedulingPassBase<ModuloSchedulingPass>
+  {
+    //ModuloScheduling(int unrollFactor):unrollFactor_(unrollFactor){}
+    ModuloSchedulingPass() = default;
+    ModuloSchedulingPass(const ModuloSchedulingPass &pass) {}
+    void getDependentDialects(DialectRegistry &registry) const override
+    {
+      registry.insert<scf::SCFDialect>();
+    }
+    void runOnFunction() override
+    {
+      // Get the current FuncOp operation being operated on.
+      auto f = getFunction();
+
+      scf::ForOp loop;
+
+      //  Unroll the kernel
+      f.walk([&](scf::ForOp forop)
+             {
+               if (forop.getNumIterOperands())
+               {
+                 loop = forop;
+               }
+             });
+
+      if (loop)
```

Comment:
Structure-wise, we would try to layer this as:

1.  a simple helper function that takes a builder and returns the elements of IR that it creates/modify so that the APIs would compose. An example/motivation/analogy is: think of tiling that returns the newly tiled loops for further composition rather than void (which would force a subsequent analysis to re-traverse the IR and infer what are the tiled loops it wants to do something on).
2. ensure that helper function can be called with a RewritePattern builder and wrap it into a pattern that we can then expose.
3. later, hook up to mechanisms that control pattern application.

Reply:
This is  a genuine question to learn about compiler development: what do you mean exactly? My idea for this pass was:
a) Look for the kernel loop
b) Apply modulo scheduling (I am writing a version to generalize to more than double buffering)
Is your suggestion to return the modified loop? Or also its prologue/epilogue ?

Reply:
This is probably a longer discussion but the basic idea is we try to apply transformations as rewrite patterns in a bunch of places so that they compose better. We specifically try to avoid that thinking derived from traditional compiler passes design brings us into corners re composition. This is related to the phase ordering problem and Cliff Click's 95 paper "Combining Analyses, Combining Optimizations". No need to go deeper for now, we'll have plenty of time to revisit when appropriate.